### ISSUE TITLE
"setData..." methods are asynchronous

### ISSUE THREAD

**Issue by jmanos** [handsontable/handsontable#6245]
I am assuming this issue is not an intended feature, so please correct me if I am wrong.

Executing one of the "getData..." methods such as `getDataAtCell` or `getDataAtRowProp` immediately after `setDataAtCell` or `setDataAtRowProp` does not return the data that was set with the "setData..." methods. Instead, the old data is returned. This unexpected behavior is not documented in the API Reference (at least not under the methods' references).

https://jsfiddle.net/aeuszjyr/

Wrapping the "getData..." methods in a `setInterval` allows the new value to be obtained (not included in the fiddle). Of course, an `afterChange` hook could be used as well.

**Comment by AMBudnik** [comment]
You're right @jmanos. We should get the new value in these cases.

By adding a `1ms` delay 

```
setTimeout(function(){
	document.querySelector("#cell").innerText = hot.getDataAtCell(0, 0);
}, 1)
```

we're already changing the order in the stack and getting the right value.

**Comment by dongapplyboard** [comment]
I'm having trouble implementing the `afterChange` hook in react with this. My goal is to change an entire column of cells like this.

```
  // set new status of all loaded applicants
  setStatus(value) {
    if (!value) return;
    const rows = this.hotTableComponent.current.hotInstance.countRows();
    const emptyRows = this.hotTableComponent.current.hotInstance.countEmptyRows();
    // only update rows with values in them
    const rowsToUpdate = rows - emptyRows;

    for (let i = 0; i < rowsToUpdate; i++) {
      this.hotTableComponent.current.hotInstance.setDataAtCell(i, 5, value);
    }
  }

  // set procssed by column
  setProcessedBy(value) {
    if (!value) return;
    const rows = this.hotTableComponent.current.hotInstance.countRows();
    const emptyRows = this.hotTableComponent.current.hotInstance.countEmptyRows();
    // only update rows with values in them
    const rowsToUpdate = rows - emptyRows;

    for (let i = 0; i < rowsToUpdate; i++) {
      this.hotTableComponent.current.hotInstance.setDataAtCell(i, 7, value);
    }
  }
```

and then get the data to `POST` to the backend with `this.hotTableComponent.current.hotInstance.getData()` but as OP stated, the old data is returned instead of the new. I also tried doing it asynchronously by adding a an `await` before `setStatus` and `setProcessedBy` but the `getData()` still returns the old data before the changes are made.

**Comment by AMBudnik** [comment]
Hey @dongapplyboard 
have you tried the `setTimeout` for checking the data?

**Comment by dongapplyboard** [comment]
> Hey @dongapplyboard
> have you tried the `setTimeout` for checking the data?

@AMBudnik  A `setTimeout` works but this shouldn't be the desired way to fix this problem since updating thousands of records at once (eg. taking maybe 5 seconds) will mean the setTimeout has to be dynamic compared to updating 2 records.

Is there another solution for waiting until all the `setDataAtCell` are finished and then getting the data from the table?

**Comment by AMBudnik** [comment]
>  this shouldn't be the desired way to fix this problem

agree, I just thought that it might not change the situation in some cases.

> Is there another solution for waiting until all the setDataAtCell are finished and then getting the data from the table?

There's a hook called [afterSetDataAtCell](https://handsontable.com/docs/7.1.1/Hooks.html#event:afterSetDataAtCell).

**Comment by dongapplyboard** [comment]
> > this shouldn't be the desired way to fix this problem
> 
> agree, I just thought that it might not change the situation in some cases.
> 
> > Is there another solution for waiting until all the setDataAtCell are finished and then getting the data from the table?
> 
> There's a hook called [afterSetDataAtCell](https://handsontable.com/docs/7.1.1/Hooks.html#event:afterSetDataAtCell).

@AMBudnik I've tried utilizing that and `handleAfterChange` however I need to do a bulk update to a huge amount of records and then send that data to an API. The handle `afterSetDataAtCell` will fire each time. I guess I can add a debounce but this will cause alot of latency problems.

**Comment by adrianspdev** [comment]
Hi all, thank you for the report and for the discussion. I re-checked this against the current version (18.0.0) and want to close the loop.

**Summary: this is expected behavior, not a bug.** The behavior is unchanged since 2013 and is now documented.

### What happens

`setDataAtCell()` and `setDataAtRowProp()` write to the data source only after every changed cell has been validated. Validation is always deferred, even when the validator itself is synchronous. That includes the built-in validators used by the `numeric`, `date`, `time`, `dropdown`, and `autocomplete` cell types, which is why the fiddle above (a `numeric` column) shows the old value right after the call.

- Cell has **no validator** (for example a plain text column): the change is applied synchronously, and `getDataAtCell()` returns the new value immediately.
- Cell has **any validator**: the change is applied in a microtask (since 15.3.0; earlier versions used a `setTimeout(…, 0)`), so `getDataAtCell()` called in the same synchronous block still returns the old value.

Verified with 18.0.0:

```js
const hot = new Handsontable(el, {
  data: [{ id: 1, name: 'a' }],
  columns: [{ data: 'id', type: 'numeric' }, { data: 'name' }],
});

hot.setDataAtCell(0, 0, 100);
hot.getDataAtCell(0, 0); // 1   (numeric -> validator -> deferred)

hot.setDataAtCell(0, 1, 'b');
hot.getDataAtCell(0, 1); // 'b' (no validator -> synchronous)
```

### How to read the value after the change

Use the `afterChange` hook. It fires exactly once per `setDataAtCell()` call, after all validators for that call have resolved and the data has been written, regardless of how many cells the call changed. That also covers the bulk-update case raised in this thread: pass all changes in one `setDataAtCell([[row, col, value], ...])` call and read the data in the single `afterChange` that follows.

```js
hot.addHook('afterChange', (changes, source) => {
  if (source === 'myUpdate') {
    sendToApi(hot.getData());
  }
});

hot.setDataAtCell(changes, 'myUpdate');
```

Note that `afterSetDataAtCell` fires *before* validation and before the data is written, so it is not the right hook for this. The API reference for that hook now says so.

### Documentation

- [How validation affects hook order](https://handsontable.com/docs/javascript-data-grid/events-and-hooks/#how-validation-affects-hook-order) in the Events and hooks guide.
- [`afterSetDataAtCell`](https://handsontable.com/docs/javascript-data-grid/api/hooks/#aftersetdataatcell) and [`afterSetDataAtRowProp`](https://handsontable.com/docs/javascript-data-grid/api/hooks/#aftersetdataatrowprop) point to `afterChange` for reacting after the write.

Since the behavior is intentional and documented, I am closing this issue. If you run into a case where `afterChange` does not fire after the validators resolve, please open a new issue with a reproduction.
